feat: add `stats/nanmeanwd` (#8786)
stdlib-js/stdlib: lib/node_modules/@stdlib/stats/nanmeanwd/README.md

<!--

@license Apache-2.0

Copyright (c) 2025 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# nanmeanwd

> > Compute the [arithmetic mean][arithmetic-mean] along one or more [ndarray][@stdlib/ndarray/ctor] dimensions using Welford's algorithm, ignoring `NaN` values.

Reviewer: ```suggestion
> Compute the [arithmetic mean][arithmetic-mean] along one or more [ndarray][@stdlib/ndarray/ctor] dimensions, ignoring `NaN` values and using Welford's algorithm.
```

Let's reword the description slightly here and elsewhere.

// returns 'float64'
```

#### nanmeanwd.assign( x, out\[, options] )

Computes the [arithmetic mean][arithmetic-mean] along one or more [ndarray][@stdlib/ndarray/ctor] dimensions using Welford's algorithm, ignoring `NaN` values, and assigns results to a provided output [ndarray][@stdlib/ndarray/ctor].

```javascript
var array = require( '@stdlib/ndarray/array' );
var zeros = require( '@stdlib/ndarray/zeros' );

var x = array( [ 1.0, NaN, -2.0, 4.0 ] );
var y = zeros( [] );

var out = nanmeanwd.assign( x, y );
// returns <ndarray>

var v = out.get();
// returns 1.0

var bool = ( out === y );
// returns true
```

The method has the following parameters:

-   **x**: input [ndarray][@stdlib/ndarray/ctor]. Must have a real-valued or generic [data type][@stdlib/ndarray/dtypes].
-   **out**: output [ndarray][@stdlib/ndarray/ctor].
-   **options**: function options (_optional_).

The method accepts the following options:

-   **dims**: list of dimensions over which to perform a reduction. If not provided, the function performs a reduction over all elements in a provided input [ndarray][@stdlib/ndarray/ctor].

</section>

<!-- /.usage -->

<section class="notes">

## Notes

> -   This function uses Welford's algorithm for computing the mean, which provides improved numerical stability compared to naive summation approaches, particularly beneficial when dealing with values of different magnitudes or accumulating floating-point errors.

Reviewer: ```suggestion
-   This function uses Welford's algorithm for computing the arithmetic mean, providing improved numerical stability compared to naive summation approaches and being particularly beneficial when dealing with values of different magnitudes or accumulating floating-point errors.
```

> -   This function uses Welford's algorithm for computing the arithmetic mean, providing improved numerical stability compared to naive summation approaches and being particularly beneficial when dealing with values of different magnitudes or accumulating floating-point errors.

Reviewer: ```suggestion
-   This implementation uses Welford's algorithm for computing the arithmetic mean, providing improved numerical stability compared to naive summation approaches and being particularly beneficial when dealing with values of different magnitudes or accumulating floating-point errors.
```

</section>

<!-- /.notes -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var uniform = require( '@stdlib/random/base/uniform' );
var filledarrayBy = require( '@stdlib/array/filled-by' );
var bernoulli = require( '@stdlib/random/base/bernoulli' );
var getDType = require( '@stdlib/ndarray/dtype' );
var ndarray2array = require( '@stdlib/ndarray/to-array' );
var ndarray = require( '@stdlib/ndarray/ctor' );
var nanmeanwd = require( '@stdlib/stats/nanmeanwd' );

function rand() {
    if ( bernoulli( 0.8 ) < 1 ) {
        return NaN;
    }
    return uniform( 0.0, 20.0 );
}

// Generate an array of random numbers:
var xbuf = filledarrayBy( 25, 'generic', rand );

// Wrap in an ndarray:
var x = new ndarray( 'generic', xbuf, [ 5, 5 ], [ 5, 1 ], 0, 'row-major' );
console.log( ndarray2array( x ) );

// Perform a reduction:
var y = nanmeanwd( x, {
    'dims': [ 0 ]
});

// Resolve the output array data type:
var dt = getDType( y );
console.log( dt );

// Print the results:
console.log( ndarray2array( y ) );
```

</section>

<!-- /.examples -->

> 

Reviewer: You should add the references for Welford's algorithm (see `stats/base/ndarray/nanmeanwd`).

See the License for the specific language governing permissions and
limitations under the License.

-->

# nanmeanwd

> Compute the [arithmetic mean][arithmetic-mean] along one or more [ndarray][@stdlib/ndarray/ctor] dimensions, ignoring `NaN` values and using Welford's algorithm.

<section class="intro">

The [arithmetic mean][arithmetic-mean] is defined as

<!-- <equation class="equation" label="eq:arithmetic_mean" align="center" raw="\mu = \frac{1}{n} \sum_{i=0}^{n-1} x_i" alt="Equation for the arithmetic mean."> -->

```math
\mu = \frac{1}{n} \sum_{i=0}^{n-1} x_i
```

<!-- <div class="equation" align="center" data-raw-text="\mu = \frac{1}{n} \sum_{i=0}^{n-1} x_i" data-equation="eq:arithmetic_mean">
    <img src="https://cdn.jsdelivr.net/gh/stdlib-js/stdlib@42d8f64d805113ab899c79c7c39d6c6bac7fe25c/lib/node_modules/@stdlib/stats/strided/nanmean/docs/img/equation_arithmetic_mean.svg" alt="Equation for the arithmetic mean.">
    <br>
</div> -->

<!-- </equation> -->

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var nanmeanwd = require( '@stdlib/stats/nanmeanwd' );
```

#### nanmeanwd( x\[, options] )

> Computes the [arithmetic mean][arithmetic-mean] along one or more [ndarray][@stdlib/ndarray/ctor] dimensions using Welford's algorithm, ignoring `NaN` values.

Reviewer: ```suggestion
Computes the [arithmetic mean][arithmetic-mean] along one or more [ndarray][@stdlib/ndarray/ctor] dimensions, ignoring `NaN` values and using Welford's algorithm.
```

    'shape': [ 2, 2 ],
    'order': 'row-major'
});

var v = ndarray2array( x );
// returns [ [ 1.0, NaN ], [ -2.0, 4.0 ] ]

var y = nanmeanwd( x, {
    'dims': [ 0 ],
    'keepdims': true
});
// returns <ndarray>

v = ndarray2array( y );
// returns [ [ -0.5, 4.0 ] ]

y = nanmeanwd( x, {
    'dims': [ 1 ],
    'keepdims': true
});
// returns <ndarray>

v = ndarray2array( y );
// returns [ [ 1.0 ], [ 1.0 ] ]

y = nanmeanwd( x, {
    'dims': [ 0, 1 ],
    'keepdims': true
});
// returns <ndarray>

v = ndarray2array( y );
// returns [ [ 1.0 ] ]
```

By default, the function returns an [ndarray][@stdlib/ndarray/ctor] having a [data type][@stdlib/ndarray/dtypes] determined by the function's output data type [policy][@stdlib/ndarray/output-dtype-policies]. To override the default behavior, set the `dtype` option.

```javascript
var getDType = require( '@stdlib/ndarray/dtype' );
var array = require( '@stdlib/ndarray/array' );

var x = array( [ 1.0, NaN, -2.0, 4.0 ], {
    'dtype': 'generic'
});

var y = nanmeanwd( x, {
    'dtype': 'float64'
});
// returns <ndarray>

var dt = String( getDType( y ) );
// returns 'float64'
```

#### nanmeanwd.assign( x, out\[, options] )

> Computes the [arithmetic mean][arithmetic-mean] along one or more [ndarray][@stdlib/ndarray/ctor] dimensions using Welford's algorithm, ignoring `NaN` values, and assigns results to a provided output [ndarray][@stdlib/ndarray/ctor].

Reviewer: ```suggestion
Computes the [arithmetic mean][arithmetic-mean] along one or more [ndarray][@stdlib/ndarray/ctor] dimensions, ignoring `NaN` values and using Welford's algorithm, and assigns results to a provided output [ndarray][@stdlib/ndarray/ctor].
```

Reviewer: By updating the description here and elsewhere, I meant throughout this PR. You seemed to have only updated the first description.

```javascript
var array = require( '@stdlib/ndarray/array' );
var zeros = require( '@stdlib/ndarray/zeros' );

var x = array( [ 1.0, NaN, -2.0, 4.0 ] );
var y = zeros( [] );

var out = nanmeanwd.assign( x, y );
// returns <ndarray>

var v = out.get();
// returns 1.0

var bool = ( out === y );
// returns true
```

The method has the following parameters:

-   **x**: input [ndarray][@stdlib/ndarray/ctor]. Must have a real-valued or generic [data type][@stdlib/ndarray/dtypes].
-   **out**: output [ndarray][@stdlib/ndarray/ctor].
-   **options**: function options (_optional_).

The method accepts the following options:

-   **dims**: list of dimensions over which to perform a reduction. If not provided, the function performs a reduction over all elements in a provided input [ndarray][@stdlib/ndarray/ctor].

</section>

<!-- /.usage -->

<section class="notes">

## Notes

> -   This implementation uses Welford's algorithm for computing the arithmetic mean, providing improved numerical stability compared to naive summation approaches and being particularly beneficial when dealing with values of different magnitudes or accumulating floating-point errors.

Reviewer: ```suggestion
-   This implementation uses Welford's algorithm for computing the arithmetic mean. Welford's algorithm can provide improved numerical stability compared to naive summation approaches and can be particularly beneficial when dealing with values of different magnitudes or the accumulation of floating-point errors.
```

> -   This implementation uses Welford's algorithm for computing the arithmetic mean. Welford's algorithm can provide improved numerical stability compared to naive summation approaches and can be particularly beneficial when dealing with values of different magnitudes or the accumulation of floating-point errors.

Reviewer: ```suggestion
-   This implementation uses Welford's algorithm for computing the arithmetic mean. Welford's algorithm can provide improved numerical stability compared to naive summation approaches and can be particularly beneficial when wanting to minimize the accumulation of floating-point errors when dealing with values of varying magnitudes.
```
